File: DSA/src/Questions/BinarySearch/TwoSumRotated.java
package Questions.BinarySearch;

import java.util.Arrays;
/*
 https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/description/

Finding the Smallest Element:
    left = 0, right = n - 1
    mid = (left + right) >> 1
    if nums[mid] > nums[right], left = mid + 1
    else right = mid

    return left

Two-Pointer Technique:
    Initialize i = pivot and j = (pivot - 1) % n.

    if nums[i] + nums[j] == target, return [nums[i], nums[j]].
    if nums[i] + nums[j] > target, decrement j.
    if nums[i] + nums[j] < target, increment i.

 */
public class TwoSumRotated {
    public static int[] twoSum(int[] nums, int target) {
        int pivot = findPivot(nums);
        System.out.println("Pivot: " + pivot); // Output: "Pivot: 4
        int left = pivot;
        int right = (pivot - 1) % nums.length;

        while (left != right) {
            System.out.println("Left: " + left + " Right: " + right); // Output: "Left: 4 Right: 3"
            int sum = nums[left] + nums[right];
            System.out.println("Sum: " + sum);
            if (sum == target) {
                return new int[] { nums[left], nums[right] };
            } else if (sum > target) {
                right = (right - 1) % nums.length;
            } else {
                left = (left + 1) % nums.length;
            }
        }

        return new int[0];
    }

    private static int findPivot(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }

    public static void main(String[] args) {
        int[] nums = { 4, 5, 6, 7, 8, 0, 1, 2 };
        int target = 9;
        int[] result = twoSum(nums, target);
        System.out.println(Arrays.toString(result)); // Output: [2, 7]
    }
}

File: DSA/Python/Questions/BinarySearch/TwoSumRotated.py
"""
# Question: Two Sum in Rotated Sorted Array
# Problem Statement:
# Given a rotated sorted array, find two numbers that sum up to a target value.
# Return the two numbers in any order.

# Example:
# Input: nums = [4,5,6,7,0,1,2], target = 6
# Output: [4,2] or [2,4]
# Explanation: 4 + 2 = 6
"""

from typing import List


class TwoSumRotated:
    def find_pair_binary(self, nums: List[int], target: int) -> List[int]:
        """
        Binary search approach
        Time Complexity: O(n log n)
        Space Complexity: O(1)
        """

        def find_pivot():
            left, right = 0, len(nums) - 1
            while left < right:
                mid = (left + right) >> 1
                if nums[mid] > nums[right]:
                    left = mid + 1
                else:
                    right = mid
            return left

        pivot = find_pivot()
        n = len(nums)
        left = pivot
        right = (pivot - 1) % n

        while left != right:
            current_sum = nums[left] + nums[right]
            if current_sum == target:
                return [nums[left], nums[right]]
            elif current_sum < target:
                left = (left + 1) % n
            else:
                right = (right - 1 + n) % n
        return []

    def find_pair_two_pointer(self, nums: List[int], target: int) -> List[int]:
        """
        Two pointer approach
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        n = len(nums)
        max_idx = 0

        # Find maximum element index
        for i in range(1, n):
            if nums[i] < nums[i - 1]:
                max_idx = i - 1
                break

        left = (max_idx + 1) % n  # Smallest element
        right = max_idx  # Largest element

        while left != right:
            curr_sum = nums[left] + nums[right]
            if curr_sum == target:
                return [nums[left], nums[right]]
            elif curr_sum < target:
                left = (left + 1) % n
            else:
                right = (right - 1 + n) % n
        return []

    def find_pair_hash(self, nums: List[int], target: int) -> List[int]:
        """
        Hash table approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        seen = set()
        for num in nums:
            complement = target - num
            if complement in seen:
                return [num, complement]
            seen.add(num)
        return []


def main():
    test_cases = [
        {"nums": [4, 5, 6, 7, 0, 1, 2], "target": 6, "expected_sum": 6},
        {"nums": [3, 4, 5, 1, 2], "target": 7, "expected_sum": 7},
        {"nums": [2, 3, 4, 5, 1], "target": 3, "expected_sum": 3},
    ]

    solution = TwoSumRotated()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Array: {test['nums']}")
        print(f"Target: {test['target']}")

        result_binary = solution.find_pair_binary(test["nums"], test["target"])
        result_two_pointer = solution.find_pair_two_pointer(
            test["nums"], test["target"]
        )
        result_hash = solution.find_pair_hash(test["nums"], test["target"])

        print(f"Binary Search Result: {result_binary}")
        print(f"Two Pointer Result: {result_two_pointer}")
        print(f"Hash Table Result: {result_hash}")

        # Verify results
        if result_binary:
            assert sum(result_binary) == test["expected_sum"]
        if result_two_pointer:
            assert sum(result_two_pointer) == test["expected_sum"]
        if result_hash:
            assert sum(result_hash) == test["expected_sum"]


if __name__ == "__main__":
    main()
